Makefile modifications required for gcc >= 4.9 (fixes #981)

hstm opened the pull request:
Fixes #981

This fix adds conditional cmake parameters that are required to build nupic.core with gcc version 4.9 (and higher) in combination with LTO (-flto). gcc >= 4.9 requires to use its own wrappers for ar, nm and ranlib instead of the regular binutils versions.

hstm commented:
For the records: Don't try to use conditional statements within ExternalProject_Add() functions.

rhyolight commented:
I want to test this with Ubuntu 15 / 16 Monday. I will also ask @vitaly-krugl and @breznak to review.

hstm commented:
Great, thank you! I hadn't have the chance to test the build in an Ubuntu 16.04./gcc 5.x environment yet, but a build on Ubuntu 15.10 amd64 (4.2.0-38-generic #45-Ubuntu SMP) with gcc version 4.9.3 performed well with all tests passed.

hstm commented:
Just tested:

**Ubuntu 16.04 amd64 fresh install w/ latest official updates (4.4.0-24-generic #43-Ubuntu SMP)
gcc version 5.3.1 20160413 (Ubuntu 5.3.1-14ubuntu2.1) 
cmake version 3.5.1**

Standard config/build procedure:
`cmake $NUPIC_CORE -DCMAKE_INSTALL_PREFIX=../release -DPY_EXTENSIONS_DIR=$NUPIC_CORE/bindings/py/nupic/bindings`
`make`
=> Build successful.

Tests
`$NUPIC_CORE/build/release/bin/cpp_region_test`
`$NUPIC_CORE/build/release/bin/unit-tests`
=> all pass.

@rhyolight If you can confirm this on Monday, this means that nupic.core now builds on gcc versions 4.9 and 5.x (and the latest Ubuntu LTS version).

breznak commented on external/Apr1Lib.cmake:
```diff
@@ -43,8 +43,35 @@ set(aprlib_cflags "${EXTERNAL_C_FLAGS_OPTIMIZED} ${COMMON_COMPILER_DEFINITIONS_S
+        set(aprlib_config_options ${aprlib_config_options} --enable-debug)
+    endif()
+
+    ExternalProject_Add(Apr1StaticLib
```
Maybe just add this (and below) under the `UNIX` IF() and add IF(gcc == 4.9)?

hstm replied:
Good idea. I'll update the fix.

breznak commented:
Great work @hstm ! We needed such patch very much for the coming OSes :+1: 
I'll be able to test later this week, but you've showed it runs well.

rhyolight commented:
I just confirmed that this works on ubuntu 15 with GCC 4.9 and 5.2, and that's good enough for a 👍  from me. Would still like for @vitaly-krugl to review, if possible.

vitaly-krugl commented:
Reviewing ... will add a comment when I'm done

vitaly-krugl commented:
Pls fix indentations - use spaces instead of tabs. There are many tabs in the PR.

vitaly-krugl commented:
@hstm, please hold off on pushing new commits to this PR until I let you know that I am completely done with this round of code review. Thx.

vitaly-krugl commented on external/CMakeLists.txt:
```diff
@@ -24,6 +24,13 @@ project(nupic_core_main CXX)
+# fixes #981
+IF(UNIX AND CMAKE_COMPILER_IS_GNUCXX AND (CMAKE_CXX_COMPILER_VERSION VERSION_GREATER "4.9" OR CMAKE_CXX_COMPILER_VERSION VERSION_EQUAL "4.9"))
+    set(CMAKE_AR "/usr/bin/gcc-ar")
+    set(CMAKE_RANLIB "/usr/bin/gcc-ranlib")
```
Both of these variables were already set in the top-level CMakeLists.txt. Shouldn't need to set here.

hstm replied:
You're right, changes in external/CMakeLists.txt are obsolete. Will change that.

vitaly-krugl commented on CMakeLists.txt:
```diff
@@ -28,6 +28,12 @@ set(CMAKE_VERBOSE_MAKEFILE OFF)
+# fixes #981
+IF(UNIX AND CMAKE_COMPILER_IS_GNUCXX AND (CMAKE_CXX_COMPILER_VERSION VERSION_GREATER "4.9" OR CMAKE_CXX_COMPILER_VERSION VERSION_EQUAL "4.9"))
+    set(CMAKE_AR "/usr/bin/gcc-ar")
+    set(CMAKE_RANLIB "/usr/bin/gcc-ranlib")
```
I think that the intention was that these variables - `CMAKE_RANLIB`  and `CMAKE_AR` would be used in other modules, but each of the modules presently uses "/usr/bin/gcc-ar" and "/usr/bin/gcc-ranlib" verbatim, which doesn't sound right.

hstm replied:
Will remove the absolute path in CMAKE_AR and CMAKE_RANLIB.
Both variables seem to be used in the config process (if you remove the variables, you'll get a lot of "lto plugin needed" errors and build fails).

vitaly-krugl replied:
The absolute paths for "/usr/bin/gcc-ar" and "/usr/bin/gcc-ranlib" may be a bad idea, since it only works if the toolchain is installed using those absolute paths versus relying on the environment. Also, your changes in Apr1Lib.cmake and AprUtil1Lib.cmake don't use absolute paths.

vitaly-krugl replied:
Since you have these variables defined at root level, are the explicit -DCMAKE_AR= and -DCMAKE_RANLIB= in external .cmake files actually needed?

vitaly-krugl commented on external/Apr1Lib.cmake:
```diff
@@ -51,24 +51,47 @@ if (UNIX)
+        	INSTALL_COMMAND
+            	make -f Makefile install
+    	)
+    else()
```
The duplications, such as this one, in all of the external modules makes it more difficult to maintain.

hstm replied:
Yeah, you're right. It would be best if you could use an if-statement within the ExternalProject_Add() function block, but that doesn't work (cmake just ignores the if-condition and silently executes the contents of the if-block no matter what the condition is).

vitaly-krugl replied:
@hstm, we can place the extra values in variables (empty by default) instead of duplicating the entire block.

vitaly-krugl commented on external/AprUtil1Lib.cmake:
```diff
@@ -47,27 +47,51 @@ if (UNIX)
+        	PATCH_COMMAND ""
+
+        	CONFIGURE_COMMAND
+            	<SOURCE_DIR>/configure AR=gcc-ar NM=gcc-nm RANLIB=gcc-ranlib
```
Is `gcc-nm` needed here and in Apr1Lib.cmake? Do the Apr builds require the ability to list symbols?

vitaly-krugl commented:
Since the new parameters are tied to LTO, and LTO optimization flags are set up in https://github.com/numenta/nupic.core/blob/master/CommonCompilerConfig.cmake, I think it's reasonable to expect that the correlated parameters should be set up in the same file for readability and maintainability. Here is what I am thinking:

1. In CommonCompilerConfig.cmake:
 1. Export  a new list variable COMMON_CMAKE_DEFINITIONS_OPTIMIZED (default to empty list) that contains `-DCMAKE_AR=/usr/bin/gcc-ar` and `-DCMAKE_RANLIB=/usr/bin/gcc-ranlib` if supported by the compiler toolchain (compiler version check, GNU compiler, etc.) and if optimizations are enabled (non-debug build). This should track the presence of lto-related flags in `optimization_flags_cc` - see https://github.com/numenta/nupic.core/blob/8b6caad44b3593e2c29b91c194cc619114eb9c6a/CommonCompilerConfig.cmake#L170 and https://github.com/numenta/nupic.core/blob/8b6caad44b3593e2c29b91c194cc619114eb9c6a/CommonCompilerConfig.cmake#L240 (here it's reset if build type is "Debug")
 1. b
2. Add `${COMMON_CMAKE_DEFINITIONS_OPTIMIZED}` to the various projects' cmake args lists similarly to how [${CAPNP_CMAKE_DEFINITIONS} is used in CapnProto](https://github.com/hstm/nupic.core/blob/8b6caad44b3593e2c29b91c194cc619114eb9c6a/external/CapnProto.cmake#L78)

vitaly-krugl commented:
Are AR and RANLIB set properly for src_lib_static_nupiccore_solo by cmake `add_library` tooling? (reference https://github.com/numenta/nupic.core/blob/eaa5fefa86c6c6c18b89e9189dcc8d1e18346713/src/CMakeLists.txt#L366)

vitaly-krugl commented:
@hstm, thank you for your work in figuring out the necessary parameters. I am done with the first round of code review. Please see my feedback.

vitaly-krugl commented:
There may be eventual merge and logic conflicts with #975.

hstm commented:
@vitaly-krugl Thanks for the review!

hstm commented:
- removed redundancies by introducing two new variables in CommonCompilerConfig.cmake
```
COMMON_STATICLIB_CMAKE_DEFINITIONS_OPTIMIZED
COMMON_STATICLIB_CONFIGURE_DEFINITIONS_OPTIMIZED_STR
```
- CMAKE_AR and CMAKE_RANLIB are now set in root level CMakeLists.txt only
- removed absolute paths

Notes: 
- Seems we need both settings in CMakeLists.txt and CommonCompilerConfig.cmake for the build to succeed, otherwise CMAKE_AR and CMAKE_RANLIB won't be propagated correctly (this also makes sure that AR and RANLIB are set properly for src_lib_static_nupiccore_solo by cmake add_library tooling).
- We don't need to check if LTO is enabled in optimization_flags_cc since nupic.core builds fine with the gcc-*-wrappers even LTO disabled.
- Successfully built and tested with gcc 4.9.x (Ubuntu 15.10) and gcc 5.x (Ubuntu 16.04)

@vitaly-krugl Could you please have a look at the changes? Thx.

vitaly-krugl commented on CommonCompilerConfig.cmake:
```diff
@@ -91,6 +91,8 @@ set(EXTERNAL_CXX_FLAGS_OPTIMIZED)
 set(EXTERNAL_LINKER_FLAGS_OPTIMIZED)
 
+set(COMMON_STATICLIB_CMAKE_DEFINITIONS_OPTIMIZED)
+set(COMMON_STATICLIB_CONFIGURE_DEFINITIONS_OPTIMIZED_STR "")
```
The file header contains documentation of each variable exported by this module. Pls document the two new variables there, too.

vitaly-krugl commented:
@hstm, I noticed that the new `COMMON_STATICLIB_CMAKE_DEFINITIONS_OPTIMIZED ` parameters are not propagated to CapnProto.cmake. Yet, I assume that nupic.core built successfully for you in the "gcc >= 4.9" environment. If so, this might confirm that setting `CMAKE_AR` and `CMAKE_RANLIB` in the root CMakeLists.txt propagates the values to all externals, making `COMMON_STATICLIB_CMAKE_DEFINITIONS_OPTIMIZED` unnecessary. As far as I remember, cmake variable values propagate from higher-level to lower-level in the hierarchy. If so, then you only need to set `CMAKE_AR` and `CMAKE_RANLIB` in CommonCompilerConfig.cmake and don't need `COMMON_STATICLIB_CMAKE_DEFINITIONS_OPTIMIZED`. However, you would still need `COMMON_STATICLIB_CONFIGURE_DEFINITIONS_OPTIMIZED_STR`, since the Apache Portable Runtime (Apr) library builds are not using cmake under the covers (and thus are not aware of `CMAKE_AR` and `CMAKE_RANLIB`), so we still need to pass `COMMON_STATICLIB_CONFIGURE_DEFINITIONS_OPTIMIZED_STR` to their `configure` scripts.

vitaly-krugl commented:
@hstm, I am done with the second round of code review. Please see my feedback. Thank you.

hstm commented:
@vitaly-krugl Thank you very much for the review.

CapnProto is built without optimizations (`EXTERNAL_CXX_FLAGS_UNOPTIMIZED`), that means with LTO disabled, so we don't have to use `COMMON_STATICLIB_CMAKE_DEFINITIONS_OPTIMIZED` there.

Just tried a build with gcc 4.9.3 and `COMMON_STATICLIB_CMAKE_DEFINITIONS_OPTIMIZED` **unset**:

```
[ 62%] Building C object CMakeFiles/yaml.dir/src/scanner.c.o
[ 75%] Building C object CMakeFiles/yaml.dir/src/dumper.c.o
[ 87%] Building C object CMakeFiles/yaml.dir/src/api.c.o
[100%] Building C object CMakeFiles/yaml.dir/src/writer.c.o
Linking C static library libyaml.a
/usr/bin/ar: CMakeFiles/yaml.dir/src/parser.c.o: plugin needed to handle lto object
/usr/bin/ar: CMakeFiles/yaml.dir/src/emitter.c.o: plugin needed to handle lto object
/usr/bin/ar: CMakeFiles/yaml.dir/src/loader.c.o: plugin needed to handle lto object
/usr/bin/ar: CMakeFiles/yaml.dir/src/reader.c.o: plugin needed to handle lto object
```
Of course ar should be gcc-ar in this case, otherwise the build fails. This indicates that `CMAKE_AR` and `CMAKE_RANLIB` set in the root CMakeLists.txt aren't propagated correctly to the externals (maybe this is a bug, see https://cmake.org/Bug/view.php?id=15547).

As a consequence, we have to keep the variables in CMakeLists.txt as well as both `COMMON_STATICLIB_CMAKE_DEFINITIONS_OPTIMIZED` and
`COMMON_STATICLIB_CONFIGURE_DEFINITIONS_OPTIMIZED_STR` in `CommonCompilerConfig.cmake`.

hstm commented:
@vitaly-krugl We could possibly add `${COMMON_STATICLIB_CMAKE_DEFINITIONS_OPTIMIZED}` to CapnProto.cmake although it's not required right now. It wouldn't hurt (we'd add just one line of code) and we'd make sure everything still builds in case somebody switches the CapnProto build from `EXTERNAL_CXX_FLAGS_UNOPTIMIZED` to `EXTERNAL_CXX_FLAGS_OPTIMIZED` (which could happen one day). What do you think?

rhyolight commented:
> We could possibly add ${COMMON_STATICLIB_CMAKE_DEFINITIONS_OPTIMIZED} to CapnProto.cmake although it's not required right now. It wouldn't hurt (we'd add just one line of code) and we'd make sure everything still builds in case somebody switches the CapnProto build from EXTERNAL_CXX_FLAGS_UNOPTIMIZED to EXTERNAL_CXX_FLAGS_OPTIMIZED (which could happen one day). What do you think?

@scottpurdy could have an opinion on this.

vitaly-krugl commented:
@hstm, regarding {quote}@vitaly-krugl We could possibly add `${COMMON_STATICLIB_CMAKE_DEFINITIONS_OPTIMIZED}` to CapnProto.cmake although it's not required right now. It wouldn't hurt (we'd add just one line of code) and we'd make sure everything still builds in case somebody switches the CapnProto build from `EXTERNAL_CXX_FLAGS_UNOPTIMIZED` to `EXTERNAL_CXX_FLAGS_OPTIMIZED` (which could happen one day). What do you think?{quote}

Let's leave `COMMON_STATICLIB_CMAKE_DEFINITIONS_OPTIMIZED` out of CapnProto. We don't want to mix unoptimized with optimized settings. Who knows what else might be added to `COMMON_STATICLIB_CMAKE_DEFINITIONS_OPTIMIZED` in the future?

vitaly-krugl commented on CommonCompilerConfig.cmake:
```diff
@@ -220,6 +229,12 @@ elseif(${CMAKE_CXX_COMPILER_ID} MATCHES "Clang")
 
+# Compatibility with gcc >= 4.9 which requires the use of gcc's own wrappers for ar and ranlib in combination with LTO
+# works also with LTO disabled
+IF(UNIX AND CMAKE_COMPILER_IS_GNUCXX AND (NOT "${CMAKE_BUILD_TYPE}" STREQUAL "Debug") AND (CMAKE_CXX_COMPILER_VERSION VERSION_GREATER "4.9" OR CMAKE_CXX_COMPILER_VERSION VERSION_EQUAL "4.9"))
```
To avoid duplication of logic, let's set `CMAKE_AR` and `CMAKE_RANLIB` here as well along with a short explanatory comment to the effect that `COMMON_STATICLIB_CMAKE_DEFINITIONS_OPTIMIZED` is a workaround for a cmake bug (including a link to https://gitlab.kitware.com/cmake/cmake/issues/15547), instead of doing it separately in the root CMakeLists.txt. Also, document `CMAKE_AR` and `CMAKE_RANLIB` alongside [CMAKE_LINKER](https://github.com/hstm/nupic.core/blob/8b6caad44b3593e2c29b91c194cc619114eb9c6a/CommonCompilerConfig.cmake#L64-L65)

vitaly-krugl commented on CommonCompilerConfig.cmake:
```diff
@@ -61,6 +61,13 @@
 #                      and shared libraries (DLLs) with optimizations that are
 #                      compatible with INTERNAL_CXX_FLAGS_OPTIMIZED
 #
+# COMMON_STATICLIB_CMAKE_DEFINITIONS_OPTIMIZED: list of DMAKE_AR and DMAKE_RANLIB flags required for gcc >= 4.9
```
The name of the variable is more generic than the provided description. Perhaps document it as: list of -D cmake definitions corresponding to `EXTERNAL_C_FLAGS_OPTIMIZED`. If you think it's useful, you may then site those details as examples.

Ditto for `COMMON_STATICLIB_CONFIGURE_DEFINITIONS_OPTIMIZED_STR`.

Also, since these are used only for building external projects due to the cmake bug https://gitlab.kitware.com/cmake/cmake/issues/15547, let's change the prefix of both of these variables from `COMMON_` to `EXTERNAL_` for consistency with other such flags here (sorry for suggesting otherwise earlier).

vitaly-krugl replied:
Note DMAKE_AR and DMAKE_RANLIB (DMAKE_ instead of CMAKE_)

vitaly-krugl commented:
@hstm, thanks for tracking down the `CMAKE_AR` bug. We're almost there. I have just a couple of simple feedback items this time.

scottpurdy commented:
@rhyolight I'll defer to @vitaly-krugl on this discussion.

hstm commented:
@vitaly-krugl Thanks for the review. I've just pushed the updated files:

- root `CMakelists.txt` cleaned up, `CMAKE_AR` and `CMAKE_RANLIB` moved to `CommonCompilerConfig.cmake`
- variables renamed to `EXTERNAL*`
- documentation updated

Successfully tested with gcc 4.9.3 and 5.3.1.

vitaly-krugl commented on CommonCompilerConfig.cmake:
```diff
@@ -55,12 +55,19 @@
+#                    
 # CMAKE_LINKER: updated, if needed; use ld.gold if available. See cmake
 #               documentation
 #
```
Pls add blurbs for `CMAKE_AR` and `CMAKE_RANLIB` in this documentation.

vitaly-krugl commented:
@hstm, LGTM after adding blurbs for `CMAKE_AR` and `CMAKE_RANLIB` in CommonCompilerConfig.cmake header (near the blurb for `CMAKE_LINKER`). Thanks!

hstm commented:
@vitaly-krugl Done.

vitaly-krugl commented:
@hstm, thank you for your effort and patience.

@rhyolight: LGTM at 7028279836f59da0a2ec0cc30f77b26541e742a5

vitaly-krugl commented:
I verified the build and tests on Ubuntu 16.04 running in VirtualBox; gcc v5.3.1, python v2.7.11+.

rhyolight commented:
Yes! Great work you all! 

![celebration dance](http://reactiongifs.me/wp-content/uploads/2014/01/jimmy-fallon-elmo-happy-dance-saturday-night-live.gif)

hstm commented:
@vitaly-krugl Once again thank you for the reviews - the patch now looks much prettier than in my first version ;-)
